kernel/rtlil.cc

```diff
@@ -203,93 +203,171 @@ const pool<IdString> &RTLIL::builtin_ff_cell_types() {
+		return decode_string() < other.decode_string();
+
+	if (std::get_if<std::string>(&backing))
+		return assert_get_str(ctx) < other.assert_get_str(ctx);
+
+	bitvectype& bv = assert_get_bits(ctx);
+	auto other_bv = other.assert_get_bits(ctx);
+
+	if (bv.size() != other_bv.size())
+		return bv.size() < other_bv.size();
+
+	for (size_t i = 0; i < bv.size(); i++)
+		if (bv[i] != other_bv[i])
+			return bv[i] < other_bv[i];
 	return false;
 }
 
 bool RTLIL::Const::operator ==(const RTLIL::Const &other) const
 {
-	return bits == other.bits;
+	const char* ctx = "operator==";
+	if (std::get_if<std::string>(&backing) != std::get_if<std::string>(&other.backing))
+		return decode_string() == other.decode_string();
+
+	if (std::get_if<std::string>(&backing))
+		return assert_get_str(ctx) == other.assert_get_str(ctx);
+
+	return assert_get_bits(ctx) == other.assert_get_bits(ctx);
 }
 
 bool RTLIL::Const::operator !=(const RTLIL::Const &other) const
 {
-	return bits != other.bits;
+	return !(*this == other);
+}
+
+std::vector<RTLIL::State>& RTLIL::Const::bits()
+{
+	bitvectorize();
+	return assert_get_bits("Const::bits()");
+}
+
+const std::vector<RTLIL::State>& RTLIL::Const::bits() const
+{
+	bitvectorize();
+	return assert_get_bits("Const::bits()");
+}
+
+
+std::vector<RTLIL::State> RTLIL::Const::to_bits() const
+{
+	if (auto bv = std::get_if<bitvectype>(&backing)) {
+		return *bv;
+	}
+	auto str = assert_get_str("Const::to_bits");
+	bitvectype b;
+	b.reserve(str.size() * 8);
+	for (int i = str.size()-1; i >= 0; i--) {
+		unsigned char ch = str[i];
+		for (int j = 0; j < 8; j++) {
+			b.push_back((ch & 1) != 0 ? State::S1 : State::S0);
+			ch = ch >> 1;
+		}
+	}
+	return b;
+}
+
+std::string RTLIL::Const::pretty_fmt() const {
+	if (std::get_if<std::string>(&backing))
+		return decode_string();
+	else
+		return std::to_string(as_int());
+}
+
+std::string RTLIL::Const::pretty_fmt_undef() const {
+	if (std::get_if<std::string>(&backing))
+		return decode_string();
+	else
+		return as_string();
 }
 
 bool RTLIL::Const::as_bool() const
 {
-	for (size_t i = 0; i < bits.size(); i++)
-		if (bits[i] == State::S1)
+	bitvectorize();
+	bitvectype& bv = assert_get_bits("Const::as_bool");
+	for (size_t i = 0; i < bv.size(); i++)
+		if (bv[i] == State::S1)
 			return true;
 	return false;
 }
 
 int RTLIL::Const::as_int(bool is_signed) const
 {
+	bitvectorize();
+	bitvectype& bv = assert_get_bits("Const::as_int");
 	int32_t ret = 0;
-	for (size_t i = 0; i < bits.size() && i < 32; i++)
-		if (bits[i] == State::S1)
+	for (size_t i = 0; i < bv.size() && i < 32; i++)
+		if (bv[i] == State::S1)
 			ret |= 1 << i;
-	if (is_signed && bits.back() == State::S1)
-		for (size_t i = bits.size(); i < 32; i++)
+	if (is_signed && bv.back() == State::S1)
+		for (size_t i = bv.size(); i < 32; i++)
 			ret |= 1 << i;
 	return ret;
 }
 
-std::string RTLIL::Const::as_string() const
+std::string RTLIL::Const::as_string(std::string any) const
```

Comment:
What's up with this `any` parameter? This looks unrelated to the backing change

Reply:
Yes, it's unrelated refactoring to adjacent code. I'm giving as_string a parameter that sets the representation for don't-care / any bits to deduplicate code that previously used the `bits()` interface on some strings. I think currently it only deduplicates it with `serialize_param_value` in `ast.cc` but there's other places where you could use this method now

Reply:
Ideally this would be a separate commit then. Also this would be easier to review if the `s/.bits/.bits()` change was a commit of its own, done upfront before we start making changes to the backing within `Const`.